driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/internal/BatchCursorFlux.java

```diff
@@ -84,7 +84,7 @@ private void recurseCursor(){
                 sink.complete();
             } else {
                 batchCursor.setBatchSize(calculateBatchSize(sink.requestedFromDownstream()));
-                Mono.from(batchCursor.nextWithSink(sink))
+                Mono.from(batchCursor.next(() -> sink.isCancelled()))
```

Comment:
With the most recent commit, I now wonder whether we should instead leave BatchCursor alone, and add a conditional here that tests for `sink.isCancelled()` (though I'm not totally sure what would go in the true branch).

What do you think?

Reply:
Its tricky as we have a guard against [!sink.isCancelled()](https://github.com/mongodb/mongo-java-driver/pull/1075/files/c23be9fd486b9eb9295a4d7d4d1dacb9c4d8f155..df97b0e55e27bd9fa67c597588619add657fc39b#diff-82d8a56e77330e2af60a4b410b30501aa7374a90a90a7e8079952e09b9f2758bR82), we also check if the `batchCursor.isClosed()` before trying to create the Mono.

So the race is from the start of `recurseCursor` to it being actioned by the `subscribe()` call.

Reply:
OK, I'm not thinking clearly.  We have to defer the check for isCancelled until later.  It doesn't matter what the value is right now (hence using a Supplier).